File: src/main/java/org/example/dp1d/MaxProdSubarr.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        curMax, curMin = 1, 1

        for n in nums:
            if n == 0:
                curMax, curMin = 1, 1
                continue
            tmp = n * curMax  # to avoid computation bug
            curMax = max(tmp, n * curMin, n)
            curMin = min(tmp, n * curMin, n)
            res = max(res, curMax)
        return res

File: src/main/java/org/example/dp1d/MaxProdSubarr.java
package org.example.dp1d;

public class MaxProdSubarr {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int currMax = 1, currMin = 1;

        for (int n: nums) {
            int temp = currMax * n;
            currMax = Math.max(Math.max(currMax * n, currMin * n), n);
            currMin = Math.min(Math.min(temp, currMin * n), n);
            res = Math.max(currMax, res);
        }

        return res;
    }

    public static void main(String[] args) {
        int[] nums = {2,3,-2,4};
        System.out.println(new MaxProdSubarr().maxProduct(nums));
    }
}
